### DISCUSSION TITLE
Does `etcd-snapshot prune` not work on S3?

### DISCUSSION THREAD

**Discussion by lindhe** [rancher/rke2#6287] [General]
I'm on `v1.28.10+rke2r1`.

In `/etc/rancher/rke2/config.yaml`, I have set `etcd-snapshot-retention: 20`. I was under the impression that this should cause periodic pruning of snapshots both on disk and on S3. But I recently discovered that I have around 700 snapshots in my S3 bucket, so that's clearly not working for me.

While [RKE2 docs](https://docs.rke2.io/backup_restore#s3-compatible-api-support) are not clear on what effect pruning should have on S3, the [K3s docs](https://docs.k3s.io/cli/etcd-snapshot#etcd-snapshot-and-restore-subcommands) indicate that it should prune on S3 too:

> These commands will perform as expected whether the etcd snapshots are stored locally or in an S3 compatible object store.

Additionally, I found https://github.com/rancher/rke2/issues/5216 that seems to confirm my belief that prune should work towards S3 as well as local disk.

At first, I found that I hadn't granted `Allow` on `s3:DeleteObject` in my bucket policy, so I assumed that was the issue. But even after granting that, it still will not prune on S3.

I am able to delete individual snapshots like this:

```console
[admin@rancher-node011 ~]$ sudo rke2 etcd-snapshot delete  on-demand-rancher-node011-1720524031
WARN[0000] Unknown flag --etcd-snapshot-schedule-cron found in config.yaml, skipping
WARN[0000] Unknown flag --tls-san found in config.yaml, skipping
INFO[0003] Snapshot on-demand-rancher-node011-1720524031 deleted.
INFO[0003] Snapshot on-demand-rancher-node011-1720524031 deleted.
```

But when running `prune`, nothing happens:

```console
[admin@rancher-node011 ~]$ sudo rke2 etcd-snapshot prune --debug
WARN[0000] Unknown flag --etcd-snapshot-schedule-cron found in config.yaml, skipping
WARN[0000] Unknown flag --tls-san found in config.yaml, skipping
```

Considering that `delete` works, and the fact that I do not get an error when running prune, it feels like my bucket policy is not the problem. But I don't know, since [it's not documented](https://github.com/rancher/rke2/issues/5216) what the bucket policy should be.

Any idea what the issue could be?

**Comment by brandond** [comment]
It will prune s3 if s3 is configured - either explicitly on the CLI, or in the config file.

A few recent releases of RKE2 were affected by https://github.com/k3s-io/k3s/issues/10292, however that has been resolved in the most recent release.

**Reply by lindhe** [reply]
Ah, interesting! That might be what's been causing issues for me. I see that [v1.28.11+rke2r1](https://github.com/rancher/rke2/releases/tag/v1.28.11%2Brke2r1) was recently released, I'll try that and see if things improve! Thanks!
